[SPARK-20265][MLlib] Improve Prefix'span pre-processing efficiency

Syrux opened the pull request:
## What changes were proposed in this pull request?

Improve PrefixSpan pre-processing efficency by preventing sequences of zero in the cleaned database. 
The efficiency gain is reflected in the following graph : https://postimg.org/image/9x6ireuvn/

## How was this patch tested?

Using MLlib's PrefixSpan existing tests and tests of my own on the 8 datasets shown in the graph. All
result obtained were stricly the same as the original implementation (without this change).
dev/run-tests was also runned, no error were found.

Author : Cyril de Vogelaere <cyril.devogelaere@gmail.com>

Syrux committed:
[SPARK-20265][MLlib] Improve Prefix'span pre-processing efficiency

srowen reviewed (commented):
I think that makes sense, though I'm not familiar with this implementation. The idea is that 0 is a delimiter and only needs to be added if something else was added to delimit? do existing tests definitely cover this case?

Syrux commented:
Yes exactly, the current implementation adds too much unnecessary delimiters. We this one line change, delimiter are only placed where needed. 

Currently there are no tests to verify if the algorithm cleans the sequences correctly. I only found that inneficiency by printing stuff around while I implemented other things on my local github. 

If you want, I can add some tests, but that will necessitate a small refector to separate the cleaning part in it's own method. Calling the current method would directly call the main algorithm ... ^^'

Two of the existing tests did cover cases where sequence of zero where left. However not at pertinent places (Integer/String type, variable-size itemsets clean a five at the end of the third sequence, leaving 2 zero instead of one). 

I can however vouch that the previous code worked just fine. Both the results of the old implementation and this one are the same. They also correspond to the results I obtained for another standalone CP based implementation. It's just that this code makes the pre-processing more efficient.

srowen commented:
Even a simplistic test of this case would give a lot more confidence that it's correct. If it means opening up a `private[spark]` method or two to make testing possible that seems reasonable. I don' think it needs significant change. Something needs to exercise this code path.

Syrux commented:
Ok, should I create a new Jira and push there the additionnal tests ?
Or is here completly fine, since it's related to the current change

Tell me, and I will get the change done asap :)

Syrux committed:
Separating cleaning methods in standalone methods + Tests for cleaning correctness

Syrux commented:
Yo Sean, I already pushed the requested changes in case it's the correct place to do so.
(I can just revert them, if not)

I added two new methods to allow tests. First a method which finds all frequent items in a database, second a method that actually clean the database using those frequent items. Although I didn't end up using the first method, the pre-processing method is now much clearer to understand. So I left the new method. Just tell me if I need to put that piece of code back.

I also added tests for multiple types of sequence database. More specifically, when there is max one item per itemset, when there can be multiple items per itemsets, and when cleaning the database empties it. They should cover all cases together.

Of course, the new implementation passes the tests perfectly, and the old one doesn't.
Every other thing remained as is.

Tell me if the way I did it was ok. I hope it's up to standards :)

SparkQA commented:
**[Test build #3651 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3651/testReport)** for PR 17575 at commit [`8e5db6a`](https://github.com/apache/spark/commit/8e5db6af95545121d379dbca83bedc23cbd5e6c0).

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -232,6 +200,68 @@ class PrefixSpan private (
+   *
+   * @return The internal repr of the inputted dataset. With properly placed zero delimiter.
+   */
+  private[fpm] def toDatabaseInternalRepr[Item: ClassTag](data : RDD[Array[Array[Item]]],
```
We generally start wrapping the args to 4-space indent here in this case. Then you can pull up the return type. Also no space before colon

Syrux replied:
Ok, the new version will fix that and the colon space.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -232,6 +200,68 @@ class PrefixSpan private (
+                                                     minCount : Long): Array[Item] = {
+
+    data.flatMap { itemsets =>
+      val uniqItems = mutable.Set.empty[Item]
```
While you're here you're welcome to write `itemsets.foreach(_.foreach(item => uniqItems += item))`

srowen replied:
or does `itemsets.foreach(set => uniqItems ++= set)` work?

Syrux replied:
Ok, changed the accolades to parenthesis. (I suppose that what you meant, correct me if I'm wrong)
Also, just by curiosity, do you know if that make any differences in performances ?

Syrux replied:
itemsets.foreach(set => uniqItems ++= set) does work. I will change it in my next commit. I will push it once I know what to do for the flag.

srowen commented on mllib/src/test/scala/org/apache/spark/mllib/fpm/PrefixSpanSuite.scala:
```diff
@@ -360,6 +360,55 @@ class PrefixSpanSuite extends SparkFunSuite with MLlibTestSparkContext {
+
+    val expected1 = Array(Array(0, 4, 0, 5, 0, 4, 0, 5, 0))
+      .map(x => x.map(y => {
+        if (y == 0) 0
```
I'd put this on one line

Syrux replied:
Ok, changing it

srowen commented on mllib/src/test/scala/org/apache/spark/mllib/fpm/PrefixSpanSuite.scala:
```diff
@@ -360,6 +360,55 @@ class PrefixSpanSuite extends SparkFunSuite with MLlibTestSparkContext {
+    val rdd3 = sc.parallelize(sequences3, 2).cache()
+
+    val cleanedSequence3 = PrefixSpan.toDatabaseInternalRepr(rdd3, itemToInt3).collect()
+    val expected3: Array[Array[Int]] = Array()
```
Nit: can this be `val expected3 = Array[Array[Int]]()`?

Syrux replied:
Yep, it can. It even avoids a useless cast.
I will push the new version asap

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -232,6 +200,68 @@ class PrefixSpan private (
+          allItems += 0
+        }
+      }
+      if (containsFreqItems) {
```
Is this the same as checking `allItems.size > 1` now, rather than maintain a flag?

Syrux replied:
Yes, but allItems is an arrayBuilder, so there is no size method.
I could do allIItems.result().size but I think the performance might be worse than a flag.

srowen replied:
I see. What about waiting to pre-pend the initial 0 until the end, only if not empty?

Syrux replied:
I am not sure about the performance of a pre-append on arrayBuilder. I will check them first. Back in a few minutes.

Syrux replied:
Apparently, prepending is impossible on an arrayBuilder. The method doesn't exist (http://www.scala-lang.org/api/2.12.0/scala/collection/mutable/ArrayBuilder.html).
I think the flag is our best bet for performance. Changing it to an arrayBuffer would be far worse since a type encapsulation would be forced on the ints it contain.

srowen replied:
OK no problem, leave it. Just riffing while we're editing the code.

srowen reviewed (changes requested):
Looking good, mostly tiny suggestions

Syrux committed:
Small corrections, done as requested

Syrux committed:
I had forgotten a four space indent, fixed it

Syrux committed:
Changing to itemsets.foreach(set => uniqItems ++= set)

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -232,6 +200,69 @@ class PrefixSpan private (
+  private[fpm] def findFrequentItems[Item: ClassTag](
+      data: RDD[Array[Array[Item]]],
+      minCount: Long):
+  Array[Item] = {
```
Wrap this onto the previous line

Syrux replied:
Ok, I will push the changes

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -232,6 +200,69 @@ class PrefixSpan private (
+      itemsets.foreach(set => uniqItems ++= set)
+      uniqItems.toIterator.map((_, 1L))
+    }.reduceByKey(_ + _).filter { case (_, count) =>
+        count >= minCount
```
Nit, but this should unindent one unit

Syrux replied:
Ok

srowen reviewed (commented):
A few more small comments about the code style but seems OK

SparkQA commented:
**[Test build #3658 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3658/testReport)** for PR 17575 at commit [`627bfe0`](https://github.com/apache/spark/commit/627bfe0b04180a9c7248df6a3af519f15261faa6).

SparkQA commented:
**[Test build #3658 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3658/testReport)** for PR 17575 at commit [`627bfe0`](https://github.com/apache/spark/commit/627bfe0b04180a9c7248df6a3af519f15261faa6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

Syrux committed:
Small changes in code style

SparkQA commented:
**[Test build #3661 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3661/testReport)** for PR 17575 at commit [`d799d46`](https://github.com/apache/spark/commit/d799d460e215c017b4385e8ecbbca8b92128096a).

SparkQA commented:
**[Test build #3661 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3661/testReport)** for PR 17575 at commit [`d799d46`](https://github.com/apache/spark/commit/d799d460e215c017b4385e8ecbbca8b92128096a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.
